### PR TITLE
Add async System.Data resultset and database schema APIs

### LINKED ISSUE #38028: Async System.Data resultset and database schema APIs
## Background and Motivation

This proposal introduces async counterparts to existing sync APIs for fetching schema information for resultset and database schema. The original proposal was developed by @manandre and @bgrainger in https://github.com/npgsql/npgsql/issues/2976.

## Proposed API

```c#
public class DbConnection
{
    public virtual Task<DataTable> GetSchemaAsync(string collectionName = null, string[] restrictions = null, CancellationToken cancellationToken = default);
}

public class DbDataReader
{
    public virtual Task<DataTable> GetSchemaTableAsync(CancellationToken cancellationToken = default);
}

public interface IDbColumnSchemaGenerator
{
    Task<ReadOnlyCollection<DbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default);
}

public static class DbDataReaderExtensions
{
    public static Task<ReadOnlyCollection<DbColumn>> GetColumnSchemaAsync(this DbDataReader reader, CancellationToken cancellationToken = default);
}
```

## Notes

* The default implementations of the above will call the appropriate sync counterpart, as is standard in ADO.NET.
* As these APIs aren't perf-sensitive, they return Task instead of ValueTask. Since they typically return a considerable amount of data, the extra allocation is unlikely to matter in any case. They also mostly return asynchronously.

/cc @David-Engel @cheenamalhotra @bgrainger @manandre @bricelam @ajcvickers @stephentoub @terrajobst @mgravell @FransBouma

### PR DESCRIPTION
**Pull request by roji**
Closes #38028

/cc @ajcvickers @stephentoub @manandre

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]
The ref and src LGTM.  Can we add tests?

File: src/libraries/System.Data.Common/src/System/Data/Common/DbConnection.cs [hunk 1]
```diff
@@ -143,21 +143,166 @@ public virtual void EnlistTransaction(System.Transactions.Transaction? transacti
...
+        /// calls may cause some providers to throw an exception.
+        /// </para>
+        /// </remarks>
         public virtual DataTable GetSchema(string collectionName, string?[] restrictionValues)
         {
             throw ADP.NotSupported();
         }
 
+        /// <summary>
+        /// This is the asynchronous version of <see cref="GetSchema()" />.
```
Comment (stephentoub) [inline src/libraries/System.Data.Common/src/System/Data/Common/DbConnection.cs:214]: All of this should be in remarks I think.  The summary should be similar to the sync version.

Comment (stephentoub) [reply src/libraries/System.Data.Common/src/System/Data/Common/DbConnection.cs:214]: (Same for the others)

Comment (roji) [reply src/libraries/System.Data.Common/src/System/Data/Common/DbConnection.cs:214]: @stephentoub you're right, the docs for all async methods currently point to the sync method and add some notes on the default implementation - not ideal. I've opened #39154 to clean this up globally in System.Data, will take care of that after this PR.

**Comment by roji** [comment]
Added full testing for the new APIs.

**Pull request merged**